pkg/kube/krt/join_test.go

```diff
@@ -206,3 +208,89 @@ func TestCollectionJoinSync(t *testing.T) {
 		{Named{"namespace", "name-static"}, NewLabeled(map[string]string{"app": "foo"}), "9.9.9.9"},
 	})
 }
+
+func TestJoinWithMergeCollection(t *testing.T) {
+	opts := testOptions(t)
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "svc",
+			Namespace: "namespace",
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{"app": "foo"},
+			Ports: []corev1.ServicePort{
+				{Name: "http", Port: 80, TargetPort: intstr.FromInt(8080)},
+			},
+			ClusterIP: "1.2.3.4",
+		},
+	}
+
+	svc2 := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "svc",
+			Namespace: "namespace",
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{"version": "v1"},
+			Ports: []corev1.ServicePort{
+				{Name: "http", Port: 80, TargetPort: intstr.FromInt(8080)},
+			},
+			ClusterIP: "1.2.3.4",
+		},
+	}
+
+	c := kube.NewFakeClient(svc)
+	services1 := krt.NewInformer[*corev1.Service](c, opts.WithName("Services")...)
+	SimpleServices := krt.NewCollection(services1, func(ctx krt.HandlerContext, o *corev1.Service) *SimpleService {
+		return &SimpleService{
+			Named:    Named{o.Namespace, o.Name},
+			Selector: o.Spec.Selector,
+		}
+	}, opts.WithName("SimpleServices")...)
+
+	c2 := kube.NewFakeClient(svc2)
+	services2 := krt.NewInformer[*corev1.Service](c2, opts.WithName("Services")...)
+	SimpleServices2 := krt.NewCollection(services2, func(ctx krt.HandlerContext, o *corev1.Service) *SimpleService {
+		return &SimpleService{
+			Named:    Named{o.Namespace, o.Name},
+			Selector: o.Spec.Selector,
+		}
+	}, opts.WithName("SimpleServices2")...)
+	c2.RunAndWait(opts.Stop())
+
+	AllServices := krt.JoinWithMergeCollection(
+		[]krt.Collection[SimpleService]{SimpleServices, SimpleServices2},
+		func(ts []SimpleService) *SimpleService {
+			if len(ts) == 0 {
+				return nil
+			}
+
+			simpleService := ts[0]
+
+			for i, t := range ts {
+				if i == 0 {
+					continue
+				}
+				// Existing labels take precedence
+				newSelector := maps.MergeCopy(t.Selector, simpleService.Selector)
+				simpleService.Selector = newSelector
+			}
+
+			return &simpleService
+		},
+		opts.With(
+			krt.WithName("AllServices"),
+		)...,
+	)
+	c.RunAndWait(opts.Stop())
+	assert.EventuallyEqual(t, func() bool {
+		return AllServices.WaitUntilSynced(opts.Stop())
+	}, true)
+
+	assert.EventuallyEqual(t, func() *SimpleService {
+		return AllServices.GetKey("namespace/svc")
+	}, &SimpleService{
+		Named:    Named{"namespace", "svc"},
+		Selector: map[string]string{"app": "foo", "version": "v1"},
+	})
+}
```

Comment:
nit: can this test illustrate update and delete as well

Reply:
Thanks a bunch for calling this out; it helped me find a couple of sneaky bug w.r.t adds: dafcc01ffdabceb913f120eb1918371000d59c0e